### PR TITLE
[Sharing dialog] Remaining "Share" dialog functionality

### LINKED ISSUE #17: [Project] Sharing dialog
Allow project manager to share with other users via email account, optionally sending an email to notify them. The manager can decide who can "Manage", "Contribute", and "View" the project. 

* Manage: Edit all project metadata, including feature types, forms, and ACLs.
* Contribute: Can add features and records or edit their own. Whether they can add/edit features or only records depends on the feature type.
* View: Read-only view of the entire project.

Finer-grained ACLs may be added in the future (e.g. who can edit which feature types.

### PR DESCRIPTION
**Pull request by gino-m**
- [x] Pending state, save button
- [x] Disallow duplicate email addresses
- [x] Edit role
- [x] Disallow editing project owner
- [x] Styling

image

@sergeydolgov1 FYI.

Closes #17.

### CODE REVIEW

File: web-ng/src/app/shared/converters/firebase-data-converter.ts [hunk 1]
```diff
@@ -456,6 +457,10 @@ export class FirebaseDataConverter {
     };
   }
 
+  static aclToJs(acl: Map<string, Role>): {} {
```
Comment (gyarill) [inline web-ng/src/app/shared/converters/firebase-data-converter.ts:460]: do you think it would be good to add test to converter?

File: web-ng/src/app/components/share-dialog/share-dialog.component.ts [hunk 2]
```diff
@@ -56,52 +77,107 @@ export class ShareDialogComponent {
     this.subscription.add(
       this.projectService
         .getActiveProject$()
+        .pipe(take(1))
```
Comment (gyarill) [inline web-ng/src/app/components/share-dialog/share-dialog.component.ts:80]: just to educate me.. why did you add it?:P

File: web-ng/src/app/components/share-dialog/share-dialog.component.ts [hunk 3]
```diff
@@ -56,52 +77,107 @@ export class ShareDialogComponent {
...
-    // TODO: Allow setting role.
-    this.projectService.updateRole(
-      this.projectId,
-      this.addUserForm.value['email'],
-      this.addUserForm.value['role']
-    );
-    // TODO: Show saving / saved status.
+    const { email, role } = this.addUserForm.value;
+
+    // Add new email/role and update change state. Validation rules prevent
```
Comment (gyarill) [inline web-ng/src/app/components/share-dialog/share-dialog.component.ts:96]: just wondering: i find it so crazy that we have only validation on client side.. firebase does not do any checks, does it? Do we care?

File: web-ng/src/app/components/share-dialog/share-dialog.component.ts [hunk 4]
```diff
@@ -56,52 +77,107 @@ export class ShareDialogComponent {
...
+
+    // Clear "Add contributor" field.
     this.addUserForm.setValue({ email: '', role: Role.CONTRIBUTOR });
   }
 
+  onRoleChange(event: MatSelectChange, index: number) {
+    if (!this.acl) {
+      return;
+    }
+    if (event.value < 0) {
```
Comment (gyarill) [inline web-ng/src/app/components/share-dialog/share-dialog.component.ts:109]: could you maybe add explanation why do we do this check?

**Comment by gyarill** [comment]
Beautiful!:D

File: web-ng/src/app/shared/converters/firebase-data-converter.ts (hunk 1 above)
Comment (gino-m) [reply web-ng/src/app/shared/converters/firebase-data-converter.ts:460]: I do! We're still punting on tests for now, but we'll add this once we get an end-to-end demo working.

File: web-ng/src/app/components/share-dialog/share-dialog.component.ts (hunk 3 above)
Comment (gino-m) [reply web-ng/src/app/components/share-dialog/share-dialog.component.ts:96]: Yes, it is, and yes we do! Firebase is schemaless, so there are no constraints on the structure/contents of data. In the future, we may implement our own microservice and respective API that will only allow valid data to be submitted.

Comment (gino-m) [reply web-ng/src/app/components/share-dialog/share-dialog.component.ts:96]: ...but for now in demo-land it's something we have to live with.

File: web-ng/src/app/components/share-dialog/share-dialog.component.ts (hunk 2 above)
Comment (gino-m) [reply web-ng/src/app/components/share-dialog/share-dialog.component.ts:83]: Added a clarifying comment. HTH!

File: web-ng/src/app/components/share-dialog/share-dialog.component.ts (hunk 3 above)
Comment (gyarill) [reply web-ng/src/app/components/share-dialog/share-dialog.component.ts:99]: nice^^

File: web-ng/src/app/components/share-dialog/share-dialog.component.ts (hunk 2 above)
Comment (gyarill) [reply web-ng/src/app/components/share-dialog/share-dialog.component.ts:83]: nice!
